spaceship/cmd/ssh.go

```diff
@@ -0,0 +1,249 @@
+	}
+	fmt.Println(status)
+	return nil
+}
+
+// ExecuteSSHSStop executes the ssh stop subcommand.
+func ExecuteSSHSStop(ctx context.Context, cmd *plugin.ExecutedCommand, chain *plugin.ChainInfo) error {
+	c, err := executeSSH(cmd, chain)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	if !c.HasRunnerScript(ctx) {
+		return ErrServerNotInitialized
+	}
+
+	stop, err := c.Stop(ctx)
+	if err != nil {
+		return err
+	}
+	fmt.Println(stop)
+	return nil
+}
+
+// ExecuteSSHRestart executes the ssh restart subcommand.
+func ExecuteSSHRestart(ctx context.Context, cmd *plugin.ExecutedCommand, chain *plugin.ChainInfo) error {
+	c, err := executeSSH(cmd, chain)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	if !c.HasRunnerScript(ctx) {
+		return ErrServerNotInitialized
+	}
+
+	restart, err := c.Restart(ctx)
+	if err != nil {
+		return err
+	}
+	fmt.Println(restart)
+	return nil
+}
+
+// ExecuteSSHLog executes the ssh log subcommand.
+func ExecuteSSHLog(ctx context.Context, cmd *plugin.ExecutedCommand, chain *plugin.ChainInfo) error {
+	c, err := executeSSH(cmd, chain)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	if !c.HasRunnerScript(ctx) {
+		return ErrServerNotInitialized
+	}
+
+	log, err := c.LatestLog()
+	if err != nil {
+		return err
+	}
+	fmt.Println(string(log))
+	return nil
+}
+
+// ExecuteSSHDeploy executes the ssh deploy subcommand.
+func ExecuteSSHDeploy(ctx context.Context, cmd *plugin.ExecutedCommand, chain *plugin.ChainInfo) error {
+	flags, err := cmd.NewFlags()
+
+	localDir, err := os.MkdirTemp(os.TempDir(), "spaceship")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = os.RemoveAll(localDir)
+	}()
+
+	var (
+		initChain, _ = flags.GetBool(flagInitChain)
+
+		localChainHome = filepath.Join(localDir, "home")
+		localBinOutput = filepath.Join(localDir, "bin")
+	)
+
+	c, err := executeSSH(cmd, chain)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	target, err := c.Target(ctx)
+	if err != nil {
+		return err
+	}
+
+	// We are using the ignite chain build command to build the app.
+	igniteChainBuildCmd := ignitecmd.NewChainBuild()
+	igniteChainBuildCmd.SetArgs([]string{
+		"-p",
+		chain.AppPath,
+		"-o",
+		localBinOutput,
+		"--release",
+		"--release.targets",
+		target,
+	})
+	if err := igniteChainBuildCmd.ExecuteContext(ctx); err != nil {
+		return err
+	}
+
+	// Extract and upload the built binary.
+	var (
+		binName           = fmt.Sprintf("%sd", chain.ChainId)
+		localChainTarball = fmt.Sprintf(
+			"%s/%s_%s.tar.gz",
+			localBinOutput,
+			chain.ChainId,
+			strings.ReplaceAll(target, ":", "_"),
+		)
+	)
+	extracted, err := tarball.Extract(ctx, localChainTarball, localBinOutput, binName)
+	if err != nil {
+		return err
+	}
+	if len(extracted) == 0 {
+		return errors.Errorf("zero files extracted from the tarball %s", localChainTarball)
+	}
+	binPath, err := c.UploadBinary(extracted[0])
+	if err != nil {
+		return err
+	}
+
+	home := c.Home()
+	if initChain || !c.HasGenesis(ctx) {
```

Comment:
We are using nohup and a script to run the chain because it is the most cross-platform way I found, but the idea is to use something like `system.d` in the future. You can deploy only on the chain for spaceship per command. In the future we can do it for more than one validator in different SSH connections using a spaceship config file or increasing the current ignite config file